### DISCUSSION TITLE
Linux port?

### DISCUSSION THREAD

**Discussion by cahsobo** [ClassicOldSong/Apollo#107] [Q&A] [answered]
Will there be a linux port?
Windows are getting crazy and crazier each update. Linux is much more reliable option nowadays. Even in gaming.

**Comment by ClassicOldSong** [comment]
https://github.com/ClassicOldSong/Apollo/issues/96

Linux is getting better at gaming due to being used commercially by Valve. But other parts are still being "free" to get better.

If you have got mainstream Sunshine working and is satisfied with it, you can try build Apollo yourself. Virtual display and clipboard sync is not supported yet but other parts should be working.

**Comment by MindfulLearner** [comment]
I had a lot of problems with NVIDIA GPUs. I've never really tried, but isn't it more viable to use Windows for gaming and WSL for coding, rather than using only Linux?

**Reply by kainatquaderee** [reply]
thanks for your opinion and no. it is not. not everyone wants to waste money on something they can get for free. so no its not more viable.

**Comment by Bammede** [comment]
linux port for major distros would be awesome

**Comment by xoniq** [comment]
I cannot wait for Linux support. I want to ditch windows for Bazzite, and SteamOS in the near future, but I'm too used to the virtual displays on all my devices, and I need that to work before I can switch.

**Reply by ClassicOldSong** [reply]
Don't be too optimistic about Linux gaming, it still has a long way to go in order to be good in extreme conditions.

Currently the only seemingly equivalent for virtual display on Linux is only doable with Wayland, but Bazzite and SteamOS are still using X. Linux being way too free in configuration is making it really hard to work with, especially to normal users who are not tech savvy enough.

**Reply by xoniq** [reply]
> Don't be too optimistic about Linux gaming, it still has a long way to go in order to be good in extreme conditions.
> 
> Currently the only seemingly equivalent for virtual display on Linux is only doable with Wayland, but Bazzite and SteamOS are still using X. Linux being way too free in configuration is making it really hard to work with, especially to normal users who are not tech savvy enough.

Even if that's just a resolution change based on the client instead of virtual monitor? I don't care much about HDR, so it's only the resolution for me.

I do use a dedicated machine with Linux just for gaming, which works amazingly well, and gone with the windows quirks.

But my stream machine is using windows still, and with the sleep issues Windows has. I let my kids wake the machine with Moonlight, the machine is in the garage with a dummy plug, and that works great, but I want to try it out with Linux too. So I guess my only option then is Sunshine with a few changes to change the resolution.

I'll experiment with it.

**Reply by ClassicOldSong** [reply]
Actually with newer hardware, Linux has way more quirks than Windows, especially in power management. This is because hardware manufacturers don't follow specs strictly and they work with Microsoft to get most of the quirks solved before made public, but for Linux you have to wait for the community to find all the quirks and adjust the code or patches, which usually is at least a 2 months delay, sometimes the issues won't be fixed in the life time of the specific hardware.

It's more of a commercial problem than a technical issue, Linux currently still takes way too less share of the market and hardware manufacturers don't put as much resources into making drivers/improving compatibility for them. Seldom have devices can get the same level of support for Linux like the Steam Deck, and the Deck gets it because Valve invest real money into it.

**Reply by ClassicOldSong** [reply]
Nvidia put more efforts on Linux drivers only because AI is becoming more and more important and it's profitable, not because gaming on Linux is becoming a thing. AMD drivers seems better but there're still unresolved bugs in MESA, we discovered, reported and they confirmed, but still not fixed for years.

**Reply by tonybarbieri** [reply]
Looks like Bazzite is using Wayland and not X11. @ClassicOldSong do you have info on the Wayland solution? I built a new PC and would love to use Bazzite but Apollo is too good and I don't want to lose it making the switch.

Edit: I saw your reply in the Linux Build discussion you linked above. Doesn't seem feasible for now so I'll stick with Windows.

**Reply by ClassicOldSong** [reply]
Gamescope in Bazzite is using Xwayland. That's an unnecessary layer of conversion...

And Gamescope doesn't perform well on Intel Iris series GPUs, giving much more graphical glitches than it should. Even on AMD GPUs there're some weird artifacts happening when using mouse but not controllers...

**Reply by eboody** [reply]
i have a virtual display and im on arch. [this](https://www.azdanov.dev/articles/2025/how-to-create-a-virtual-display-for-sunshine-on-arch-linux) guide explains it well!

**Reply by ClassicOldSong** [reply]
Thank for the info, but it only works with KDE and Xorg

**Reply by eboody** [reply]
I'm using hyprland

**Reply by ClassicOldSong** [reply]
Ah ok, but it's a persistent virtual display instead of a on-demand one. Apollo generates the EDID on the fly with requested resolution.

**Reply by eboody** [reply]
yeah youre right. although, with sunshine you can run do/undo commands so i enable/disable the displays as needed.

**Reply by eboody** [reply]
Apollo generating an EDID on the fly is so cool

**Comment by Jwbeev** [comment]
PLEASE release a Linux version soon. I too am holding out on migrating to Linux because of Apollo. As soon as Apollo supports Linux, I am making the switch. I can't stand Windows any longer!

**Reply by ClassicOldSong** [reply, marked as answer]
There will be Linux build soon, but virtual display won't be that soon. Detail: https://github.com/ClassicOldSong/Apollo/issues/96#issuecomment-2709591725

**Reply by xoniq** [reply]
> There will be Linux build soon, but virtual display won't be that soon. Detail: https://github.com/ClassicOldSong/Apollo/issues/96#issuecomment-2709591725

Understandable. Isn't it possible to alter the resolution to the connecting client? For me that is just enough. Sometimes on a steam deck, sometimes on a 4K tv. I have it headless with a 4K HDMI dongle, so the 'monitor' is capable of 4K.

Not being ignorant or anything, but I found something about using something like;

xrandr --output HDMI-1 --mode 1920x1080

To change the resolution of the 'monitor'.

Are there downsides of that usage?

**Reply by ClassicOldSong** [reply]
The other problem for Linux is, there're too many choices to make, and supporting Linux means supporting all different combinations people will make. For X server, it is the xrandr command, but for Wayland it's another.

**Reply by arjunshinoj** [reply]
> The other problem for Linux is, there're too many choices to make, and supporting Linux means supporting all different combinations people will make. For X server, it is the xrandr command, but for Wayland it's another.

X11 is almost not preferred or even used on mainstream distros now. Only specific users with extremely niche old software or ones that have a personal vendetta against wayland use x11 now. So it's better to try to focus on wayland and not bother with x for initial build.

**Reply by HitWithSpader** [reply]
> There will be Linux build soon, but virtual display won't be that soon. Detail: [#96 (comment)](https://github.com/ClassicOldSong/Apollo/issues/96#issuecomment-2709591725)

I recently found out about Apollo and I'm very intrigued. I like many of the added features. How soon would Linux builds be (without virtual display of course)  is "soon"?  I look forward to try out Apollo when a Linux gets officially released.

Currently, I'm using Sunshine and already looking to switch to Artemis on my Android devices.

**Reply by ClassicOldSong** [reply]
Apollo is buildable on Linux, you can compile yourself at the moment.

**Reply by alexcrow1974** [reply]
Still having issues with fresh built boost 1.88 installed in /usr/local, any hints?

`CMake Error: File /root/Apollo/packaging/linux/com.SudoMaker.dev.Apollo.desktop does not exist.
CMake Error at cmake/prep/special_package_configuration.cmake:20 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:43 (include)


CMake Error: File /root/Apollo/packaging/linux/com.SudoMaker.dev.Apollo.terminal.desktop does not exist.
CMake Error at cmake/prep/special_package_configuration.cmake:21 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:43 (include)


CMake Error: File /root/Apollo/packaging/linux/com.SudoMaker.dev.Apollo.metainfo.xml does not exist.
CMake Error at cmake/prep/special_package_configuration.cmake:25 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:43 (include)


CMake Warning (dev) at cmake/macros/common.cmake:27 (_find_package):
  Policy CMP0144 is not set: find_package uses upper-case <PACKAGENAME>_ROOT
  variables.  Run "cmake --help-policy CMP0144" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Environment variable BOOST_ROOT is set to:

    /usr/local/

  For compatibility, find_package is ignoring the variable, but code in a
  .cmake module might still use it.
Call Stack (most recent call first):
  cmake/dependencies/Boost_Sunshine.cmake:33 (find_package)
  cmake/dependencies/common.cmake:5 (include)
  CMakeLists.txt:57 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning at cmake/macros/common.cmake:27 (_find_package):
  Could not find a configuration file for package "Boost" that exactly
  matches requested version "1.88.0".

  The following configuration files were considered but not accepted:

    /usr/lib/x86_64-linux-gnu/cmake/Boost-1.83.0/BoostConfig.cmake, version: 1.83.0
    /lib/x86_64-linux-gnu/cmake/Boost-1.83.0/BoostConfig.cmake, version: 1.83.0

Call Stack (most recent call first):
  cmake/dependencies/Boost_Sunshine.cmake:33 (find_package)
  cmake/dependencies/common.cmake:5 (include)
  CMakeLists.txt:57 (include)


-- Boost v1.88.0 package not found in the system. Falling back to FetchContent.
-- Boost: Release build, static libraries, MPI OFF, Python OFF, testing OFF
-- Boost: libraries included: filesystem;locale;log;program_options;system;asio;crc;format;process;property_tree
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Performing Test BOOST_ATOMIC_TARGET_X86
-- Performing Test BOOST_ATOMIC_TARGET_X86 - Success
-- Performing Test BOOST_ATOMIC_COMPILER_HAS_SSE2
-- Performing Test BOOST_ATOMIC_COMPILER_HAS_SSE2 - Success
-- Performing Test BOOST_ATOMIC_COMPILER_HAS_SSE41
-- Performing Test BOOST_ATOMIC_COMPILER_HAS_SSE41 - Success
-- Performing Test BOOST_CHARCONV_QUADMATH_FOUND
-- Performing Test BOOST_CHARCONV_QUADMATH_FOUND - Success
-- Boost.Charconv: quadmath support ON
-- Boost.Context: architecture x86_64, binary format elf, ABI sysv, assembler gas, suffix .S, implementation fcontext
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Performing Test BOOST_FILESYSTEM_HAS_INIT_PRIORITY
-- Performing Test BOOST_FILESYSTEM_HAS_INIT_PRIORITY - Success
-- Performing Test BOOST_FILESYSTEM_HAS_CXX20_ATOMIC_REF
-- Performing Test BOOST_FILESYSTEM_HAS_CXX20_ATOMIC_REF - Failed
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_BLKSIZE
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_BLKSIZE - Success
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_MTIM
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_MTIM - Success
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_MTIMENSEC
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_MTIMENSEC - Failed
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_MTIMESPEC
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_MTIMESPEC - Failed
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_BIRTHTIMENSEC
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_BIRTHTIMENSEC - Failed
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_BIRTHTIMESPEC
-- Performing Test BOOST_FILESYSTEM_HAS_STAT_ST_BIRTHTIMESPEC - Failed
-- Performing Test BOOST_FILESYSTEM_HAS_STATX
-- Performing Test BOOST_FILESYSTEM_HAS_STATX - Success
-- Performing Test BOOST_FILESYSTEM_HAS_FDOPENDIR_NOFOLLOW
-- Performing Test BOOST_FILESYSTEM_HAS_FDOPENDIR_NOFOLLOW - Success
-- Performing Test BOOST_FILESYSTEM_HAS_DIRENT_D_TYPE
-- Performing Test BOOST_FILESYSTEM_HAS_DIRENT_D_TYPE - Success
-- Performing Test BOOST_FILESYSTEM_HAS_POSIX_AT_APIS
-- Performing Test BOOST_FILESYSTEM_HAS_POSIX_AT_APIS - Success
-- Performing Test BOOST_FILESYSTEM_HAS_FALLOCATE
-- Performing Test BOOST_FILESYSTEM_HAS_FALLOCATE - Success
-- Performing Test Iconv_IS_BUILT_IN
-- Performing Test Iconv_IS_BUILT_IN - Success
-- Found Iconv: built in to C library  
-- Found the following ICU libraries:
--   data (required): /usr/lib/x86_64-linux-gnu/libicudata.so
--   i18n (required): /usr/lib/x86_64-linux-gnu/libicui18n.so
--   uc (required): /usr/lib/x86_64-linux-gnu/libicuuc.so
-- Found ICU: /usr/include (found suitable version "74.2", minimum required is "4.8.1") 
-- Boost.Locale: iconv ON, ICU ON, POSIX ON, std ON, winapi OFF
-- Performing Test BOOST_LOG_TARGET_X86
`

**Reply by ClassicOldSong** [reply]
These are different errors now.

```
CMake Error: File /root/Apollo/packaging/linux/com.SudoMaker.dev.Apollo.desktop does not exist.
CMake Error at cmake/prep/special_package_configuration.cmake:20 (configure_file):
configure_file Problem configuring file
Call Stack (most recent call first):
CMakeLists.txt:43 (include)

CMake Error: File /root/Apollo/packaging/linux/com.SudoMaker.dev.Apollo.terminal.desktop does not exist.
CMake Error at cmake/prep/special_package_configuration.cmake:21 (configure_file):
configure_file Problem configuring file
Call Stack (most recent call first):
CMakeLists.txt:43 (include)

CMake Error: File /root/Apollo/packaging/linux/com.SudoMaker.dev.Apollo.metainfo.xml does not exist.
CMake Error at cmake/prep/special_package_configuration.cmake:25 (configure_file):
configure_file Problem configuring file
Call Stack (most recent call first):
CMakeLists.txt:43 (include)
```

Rename the `.desktop` `.xml` files in `packaging/linux` to the files cmake trying to find should work.

**Reply by alexcrow1974** [reply]
OK, thanks very much! I'll try this and get back to you.

**Reply by alexcrow1974** [reply]
-- Build files have been written to: /root/Apollo/build
File 'build/dev.lizardbyte.app.Sunshine.metainfo.xml' does not exist.
✘ Validation failed:

So I try:
cp com.SudoMaker.dev.Apollo.metainfo.xml dev.lizardbyte.app.Sunshine.metainfo.xml

But then end up with:

E: com.SudoMaker.dev.Apollo:3: cid-domain-not-lowercase com.SudoMaker.dev.Apollo
I: com.SudoMaker.dev.Apollo:26: description-first-word-not-capitalized

✘ Validation failed: errors: 1, infos: 1, pedantic: 1

**Reply by ClassicOldSong** [reply]
Undo other changes and change the PROJECT_FQDN in CMakeList.txt in `dev.lizardbyte.app.Sunshine` for now...

They hardcoded too much things.

**Reply by alexcrow1974** [reply]
Thanks!

**Reply by TheEldritchVoid** [reply]
I've seen a few written tutorials for how to enable headless/spoof headless on Linux most of them use/recommended Wayland and bazzite for a high % of users setup if you did do a release for Linux I'd be interested in a YouTube tutorial (I can help with that part if you can walk me through how to do it initially) just to make sure I'm going to the right places and making the right folders etc but a bit reason I'm waiting to fill send it on bazzite is as others have mentioned I'm waiting on Apollo to be supported same as sunshine. Worst case a tutorial on how to take your source and recompile for Wayland and bazzite would be another possible solve for everyone's issue 


Side question how far does Apollo differ from sunshine og now?

**Reply by ClassicOldSong** [reply]
Linux still can't do what Apollo currently does on Windows. It's not only creating a virtual display, but also matching your client's display parameters for the virtual display on the fly, while Linux can't really do the same, or they're implemented differently for different desktop environments, making things hard to manage.

Apollo currently has more features than Sunshine, including permission system, more detailed client management, better app management, clipboard sync, pause/resume scripts etc. There's a planned GUI upgrade which makes the config panel a native window instead of relying on browsers.

**Comment by MindfulLearner** [comment]
ARCHlinux lives!!!

**Reply by ClassicOldSong** [reply]
How is this related to a specific distro?

**Comment by shawny43** [comment]
Arch aur could be a game changer since Apollo can be build automatically and keep updated from git with PKGBUILD from aur packages. You can look at Arch Wiki to know more about AUR Package. Anyways, hope this could came soon to our devices.

**Reply by fxghesa** [reply]
i found this on aur [apollo-git](https://aur.archlinux.org/packages/apollo-git). but not sure if this is official

**Reply by xoniq** [reply]
> i found this on aur [apollo-git](https://aur.archlinux.org/packages/apollo-git). but not sure if this is official

No idea but it uses this git for upstream so it'll be okay.

**Comment by Lattematcha** [comment]
If I would port Apollo to Linux, I would focus solely on distributions on IDE that support virtual desktops. For example you can easily make virtual desktop on KDE through screen share portal (Which you can do through OBS or Discord), but sunshine only wants to detect physical screens so it's a shame.

**Reply by shawny43** [reply]
True thing. Last time i was used remote desktop, it's when i started the remote by Google Chrome and instead of cast an screen it's create an virtual screen suddenly on X sessions by accident. Has i know, @ClassicOldSong could relay to X Servers instead of writing a new third-party code (Windows like) to create virtual desktops. Both Xorg and Wayland support this features natively. It's just require a little tweaks for it.

**Comment by Bureaucromancer** [comment]
Has there been any thought about an approach that would treat the Linux implementation as a wholly separate session?

**Reply by Bureaucromancer** [reply]
I'm not deeply familiar with that, but looking into it... It does seem a similar concept. But in short I'd be imagining the server to create a distinct headless user session for the streaming.

Edit: on some digging I'm thinking we'd probably be talking about integrating Wolf into the server.

**Comment by HamonTeacher** [comment]
We need Linux app... Asap.

**Comment by Tausif-ZeaREX** [comment]
> Will there be a linux port?
> 
> Windows are getting crazy and crazier each update. Linux is much more reliable option nowadays. Even in gaming.
> 
> 

Yeah 
 not only that also I saw some apollo's contributor also linux lover

**Comment by EnriqueWood** [comment]
I opened [this](https://github.com/ClassicOldSong/Apollo/discussions/1516) discussion to talk about this. A achieved some sort of success on this topic on linux under KDE wayland creating virtual monitors with custom resolutions and frequencies...more details in the POC in the discussion... I hope @ClassicOldSong gets to consider this, this is as far as I can go with my current knoledge, that's why I'm asking for some help to get this implemented on Apollo
